src/platform/datapath_raw.c

```diff
@@ -589,32 +589,63 @@ CXPLAT_THREAD_CALLBACK(CxPlatRouteResolutionWorkerThread, Context)
         CxPlatDispatchLockRelease(&Worker->Lock);
 
         while (!CxPlatListIsEmpty(&Operations)) {
+            MIB_IPNET_ROW2 IpnetRow = {0};
             CXPLAT_ROUTE_RESOLUTION_OPERATION* Operation =
                 CXPLAT_CONTAINING_RECORD(
                     CxPlatListRemoveHead(&Operations), CXPLAT_ROUTE_RESOLUTION_OPERATION, WorkerLink);
-            NETIO_STATUS Status =
-            Status = GetIpNetEntry2(&Operation->IpnetRow);
-            if (Status != ERROR_SUCCESS || Operation->IpnetRow.State <= NlnsIncomplete) {
-                Status =
-                    ResolveIpNetEntry2(&Operation->IpnetRow, NULL);
-                if (Status != 0) {
+            CXPLAT_DBG_ASSERT(
+                Operation->Route.State == RouteRefreshing || Operation->Route.State == RouteResolving);
+            CXPLAT_ROUTE NewRoute = Operation->Route;
+            QUIC_STATUS Status = CxPlatQueryRoute(Operation->Socket, &NewRoute, &IpnetRow);
```

Comment:
Can you remind me why we now also have a call of GetBestRoute2 here in the worker thread? Why can't the call be just in CxPlatResolveRoute? I thought that was what we had discussed.

Reply:
In general, the ResolveIpNetEntry2 call (which is really the only reason we are queuing to a work item at all) is the "rare" case- because in the common single-gateway case, only the first connection will need to call this, and then every other connection will use the cached result (and because the persistent congestion case is really also a "rare" case IMO). So I think any duplicate calls in the worker thread are not a major perf consideration.

Reply:
I don't really care about perf at all for both cases actually. I just want to avoid having "checking if Xxx is the same" in both ResolveRoute and worker thread. I'd rather make everything async- no inline resolution at all.

Reply:
Not a perf consideration. I try to avoid duplicated code in both places.

Reply:
Fixed as we discussed offline.